main.go

```diff
@@ -68,11 +54,73 @@ outer:
 		case <-dhcpEvents:
 			debounced.Reset(cfg.DebounceInterval)
 		case <-debounced.C:
-			l, err := leases.ReadLeases(cfg.LeaseFile)
+			ls, err := leases.ReadLeases(cfg.LeaseFile)
 			if err != nil {
 				log.Fatalw("could not parse leases file", "error", err)
 			}
-			err = r.Report(l)
+			active := ls.FilterActive()
+			byMac := active.LatestByMac()
+			log.Infow("reporting leases to metal-api", "all", len(ls), "active", len(active), "uniqueActive", len(byMac))
+
+			mtx := new(sync.Mutex)
+			var items []*leases.ReportItem
+
+			wg := new(sync.WaitGroup)
+			wg.Add(len(byMac))
+
+			for mac, l := range byMac {
+				mac := mac
+				l := l
+				go func() {
+					defer wg.Done()
+
+					bmcVersion := ""
+					biosVersion := ""
+					var fru *models.V1MachineFru
+
+					ob, err := connect.OutBand(l.Ip, cfg.IpmiPort, cfg.IpmiUser, cfg.IpmiPassword, halzap.New(r.Log))
+					if err != nil {
+						log.Errorw("could not establish outband connection to device bmc", "mac", mac, "ip", l.Ip, "err", err)
+					} else {
+						bmcDetails, err := ob.BMCConnection().BMC()
+						if err != nil {
+							log.Errorw("could not retrieve bmc details of device", "mac", mac, "ip", l.Ip, "err", err)
+						} else {
+							bmcVersion = bmcDetails.FirmwareRevision
+							fru = &models.V1MachineFru{
+								BoardMfg:            bmcDetails.BoardMfg,
+								BoardMfgSerial:      bmcDetails.BoardMfgSerial,
+								BoardPartNumber:     bmcDetails.BoardPartNumber,
+								ChassisPartNumber:   bmcDetails.ChassisPartNumber,
+								ChassisPartSerial:   bmcDetails.ChassisPartSerial,
+								ProductManufacturer: bmcDetails.ProductManufacturer,
+								ProductPartNumber:   bmcDetails.ProductPartNumber,
+								ProductSerial:       bmcDetails.ProductSerial,
+							}
+						}
+
+						board := ob.Board()
+						if board != nil {
+							biosVersion = board.BiosVersion
+						}
+					}
+
```

Comment:
This is quite deeply nested, can we do something nicer ?

Reply:
Done